core/src/main/java/org/elasticsearch/discovery/zen/ZenDiscovery.java

```diff
@@ -682,171 +670,90 @@ public void clusterStateProcessed(String source, ClusterState oldState, ClusterS
         });
     }
 
-    static class ProcessClusterState {
-        final ClusterState clusterState;
-        volatile boolean processed;
-
-        ProcessClusterState(ClusterState clusterState) {
-            this.clusterState = clusterState;
-        }
-    }
+    void processNextPendingClusterState(String reason) {
+        clusterService.submitStateUpdateTask("zen-disco-receive(from master [" + reason + "])", Priority.URGENT, new ProcessedClusterStateNonMasterUpdateTask() {
```

Comment:
So, completely separate change from this PR, but how do you feel about renaming all of these log messages from "zen-disco-whatever" to either "zen-whatever" or "zend-whatever"? **Every** time I read "zen-disco" I think "zen-disconnect" instead of "zen-discovery" because they share the same prefix.